### PR TITLE
[Feature] Support hashtag in token filter type_table rules parser

### LINKED ISSUE #6867: [BUG] "# => ALPHANUM" in word_delimiter filter's type_table can cause exception.
**Describe the bug**

If `# => ALPHANUM` is the only entry in a `word_delimiter` filter's `type_table`, analysis using that filter causes an exception. If the `#` is replaced with `\u0023` then it works as expected. The same behavior is seen with the `word_delimiter_graph` filter.

**To Reproduce**
```
curl "http://localhost:9200/_analyze?pretty" \
    -s \
    -X POST \
    -H 'Content-Type: application/json' \
    -d'
{
    "tokenizer": "whitespace",
    "filter": [
        {
            "type": "word_delimiter",
            "split_on_numerics": false,
            "type_table": [
                "# => ALPHANUM"
            ]
        }
    ],
    "text": "text1 #text2"
}
'
```

OS 2.6.0 (and 2.5.0 where I originally noticed) produces:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "no_such_element_exception",
        "reason" : null
      }
    ],
    "type" : "no_such_element_exception",
    "reason" : null
  },
  "status" : 500
}
```

Server log:
```
opensearch-node  | [2023-03-29T15:20:30,600][WARN ][r.suppressed             ] [opensearch-docker-node] path: /_analyze, params: {pretty=}
opensearch-node  | org.opensearch.transport.RemoteTransportException: [opensearch-docker-node][172.31.0.2:9300][indices:admin/analyze[s]]
opensearch-node  | Caused by: java.util.NoSuchElementException
opensearch-node  | 	at java.util.TreeMap.key(TreeMap.java:1602) ~[?:?]
opensearch-node  | 	at java.util.TreeMap.lastKey(TreeMap.java:298) ~[?:?]
opensearch-node  | 	at org.opensearch.analysis.common.WordDelimiterTokenFilterFactory.parseTypes(WordDelimiterTokenFilterFactory.java:156) ~[?:?]
opensearch-node  | 	at org.opensearch.analysis.common.WordDelimiterTokenFilterFactory.<init>(WordDelimiterTokenFilterFactory.java:89) ~[?:?]
opensearch-node  | 	at org.opensearch.index.analysis.AnalysisRegistry.getComponentFactory(AnalysisRegistry.java:156) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.index.analysis.AnalysisRegistry.buildCustomAnalyzer(AnalysisRegistry.java:285) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.buildCustomAnalyzer(TransportAnalyzeAction.java:238) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.analyze(TransportAnalyzeAction.java:165) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.shardOperation(TransportAnalyzeAction.java:151) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.shardOperation(TransportAnalyzeAction.java:88) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.support.single.shard.TransportSingleShardAction.lambda$asyncShardOperation$0(TransportSingleShardAction.java:131) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:73) [opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:88) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:806) [opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
opensearch-node  | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
opensearch-node  | 	at java.lang.Thread.run(Thread.java:833) [?:?]
```

Notes:

1. This only happens if `# => ALPHANUM` is the only entry in the `type_table`. Adding a second entry causes the analysis to succeed, but it behaves as if the `# => ALPHANUM` is not present. For example:
```
curl "http://localhost:9200/_analyze?pretty" \
    -s \
    -X POST \
    -H 'Content-Type: application/json' \
    -d'
{
    "tokenizer": "whitespace",
    "filter": [
        {
            "type": "word_delimiter",
            "split_on_numerics": false,
            "type_table": [
                "# => ALPHANUM".
                "@ => ALPHANUM"
            ]
        }
    ],
    "text": "text1 #text2"
}
'
```

Produces (incorrect):
```
{
  "tokens" : [
    {
      "token" : "text1",
      "start_offset" : 0,
      "end_offset" : 5,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "text2",
      "start_offset" : 7,
      "end_offset" : 12,
      "type" : "word",
      "position" : 1
    }
  ]
}
```

Note the `text2` token should be `#text2`. The ordering of the the two `type_table` entries does not matter.

2. Changing `# => ALPHANUM` to `\\u0023 => ALPHANUM` (0x23 is of course the character code for `#`) causes the analysis to work as expected. This:
```
curl "http://localhost:9200/_analyze?pretty" \
    -s \
    -X POST \
    -H 'Content-Type: application/json' \
    -d'
{
    "tokenizer": "whitespace",
    "filter": [
        {
            "type": "word_delimiter",
            "split_on_numerics": false,
            "type_table": [
                "\\u0023 => ALPHANUM"
            ]
        }
    ],
    "text": "text1 #text2"
}
'
```

Produces (correct):
```
{
  "tokens" : [
    {
      "token" : "text1",
      "start_offset" : 0,
      "end_offset" : 5,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "#text2",
      "start_offset" : 6,
      "end_offset" : 12,
      "type" : "word",
      "position" : 1
    }
  ]
}
```

**Expected behavior**

`# => ALPHANUM` in a `word_delimiter` or `word_delimiter_graph` filter's `type_table` should treat `#` as an alphanumeric character.

**Plugins**

None.

**Screenshots**

N/A

**Host/Environment (please complete the following information):**

Official OS 2.6.0 Docker image run under Docker Desktop 4.17.0 (99724) on Mac OS 13.2.1 (Intel).

**Additional context**

None

### PR DESCRIPTION
**Pull request by noCharger**
### Description
Support hashtag in token filter type_table rules parser

### Related Issues
Resolves #6867

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/analysis/Analysis.java [hunk 1]
```diff
@@ -98,6 +99,9 @@
 public class Analysis {
     private static final Logger LOGGER = LogManager.getLogger(Analysis.class);
 
+    // Regular expression to support hashtag tokenization
+    private static Pattern hashTagRulePattern = Pattern.compile("^\\s*#\\s*=>");
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/analysis/Analysis.java:103]:
```suggestion
    private static final Pattern HASH_TAG_RULE_PATTERN = Pattern.compile("^\\s*#\\s*=>");
```

File: server/src/main/java/org/opensearch/index/analysis/Analysis.java [hunk 2]
```diff
@@ -266,7 +249,7 @@ public static <T> List<T> parseWordList(
         int lineNum = 0;
         for (String word : words) {
             lineNum++;
-            if (removeComments == false || word.startsWith("#") == false) {
+            if (word.startsWith("#") == false || hashTagRulePattern.matcher(word).find()) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/analysis/Analysis.java:252]:
```suggestion
            if (word.startsWith("#") == false || hashTagRulePattern.matcher(word).find() == true) {
```

File: modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/40_token_filters.yml [hunk 3]
```diff
@@ -127,6 +127,67 @@
...
+    - do:
+        indices.analyze:
+          body:
+            text: 'text1 #text2'
+            tokenizer: whitespace
+            filter:
+              - type: word_delimiter
+                split_on_numerics: false
+                type_table:
+                  - "# => ALPHANUM"
```
Comment (reta) [inline modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/40_token_filters.yml:159]: Could we add some tests (or update existing ones) with comments?

```
           type_table:
                  - "# This is a comment"
                  - "# => ALPHANUM"
```

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@msfroh mind taking a look please ? thank you!

File: server/src/main/java/org/opensearch/index/analysis/Analysis.java [hunk 4]
```diff
@@ -266,7 +249,7 @@ public static <T> List<T> parseWordList(
         int lineNum = 0;
         for (String word : words) {
             lineNum++;
-            if (removeComments == false || word.startsWith("#") == false) {
+            if (word.startsWith("#") == false || HASH_TAG_RULE_PATTERN.matcher(word).find() == true) {
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/index/analysis/Analysis.java:252]: Looks good!

**Review by msfroh — APPROVED** [review]

**Pull request merged**